kinesis/kinesis.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//  http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package kinesis containers the OutputPlugin which sends log records to Kinesis Stream
+package kinesis
+
+import (
+    "fmt"
+    "os"
+    "time"
+    "math/rand"
+
+    "github.com/aws/amazon-kinesis-firehose-for-fluent-bit/plugins"
+    "github.com/aws/aws-sdk-go/aws"
+    "github.com/aws/aws-sdk-go/aws/awserr"
+    "github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+    "github.com/aws/aws-sdk-go/aws/endpoints"
+    "github.com/aws/aws-sdk-go/aws/session"
+    "github.com/aws/aws-sdk-go/service/kinesis"
+    fluentbit "github.com/fluent/fluent-bit-go/output"
+    jsoniter "github.com/json-iterator/go"
+    "github.com/sirupsen/logrus"
+)
+
+const (
+    partitionKeyCharset = "abcdefghijklmnopqrstuvwxyz" + "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+)
+
+const (
+    // Kinesis API Limit https://docs.aws.amazon.com/sdk-for-go/api/service/kinesis/#Kinesis.PutRecords
+    maximumRecordsPerPut      = 500
+    maximumPutRecordBatchSize = 5242880 // 5 MB
+    maximumRecordSize         = 1048576// 1 MB
+)
+
+// PutRecordsClient contains the kinesis PutRecords method call
+type PutRecordsClient interface{
+    PutRecords(input *kinesis.PutRecordsInput) (*kinesis.PutRecordsOutput, error)
+}
+
+type random struct {
+    seededRandom    *rand.Rand
+    buffer          []byte
+}
+
+// OutputPlugin sends log records to kinesis
+type OutputPlugin struct {
+    region                          string
+    stream                          string
+    dataKeys                        string
+    partitionKey                    string
+    lastInvalidPartitionKeyIndex    int
```

Comment:
I'm confused on what this field actually does. Seems like we just use it to log info messages?

Reply:
To prvide the error message with proper description-- for which exact data record the partition key was invalid. It will help the customer to debug faster.